rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,444 @@
+      tgz-blake2-256: '...'
+      deb-blake2-256: '...'
+  - tags: [ent]
+    version: '100.0.1'
+    arch: amd64
+    flavor: ent
+    os: linux
+    stable: true
+    security_patch: true
+    deprecated: false
+    sums:
+      rpm-blake2-256: '...'
+      tgz-blake2-256: '...'
+      deb-blake2-256: '...'
+  - tags: [ent,fips]
+    version: '100.0.1'
+    arch: amd64
+    os: linux
+    flavor: ent
+    stable: true
+    security_patch: true
+    deprecated: false
+    sums:
+      rpm-blake2-256: '...'
+      tgz-blake2-256: '...'
+      deb-blake2-256: '...'
+```
+
+*note*: Since writing the above I'm leaning toward grouping targets by version number with a single set of user-assignable
+tags/labels stored at the top-level of the version entry.  Will circle back and update.
+
+#### Upgrade Policies
+
+An upgrade policy will be responsible for matching releases to servers.  In the interest of
+reliability and fault-taulerance, many upgrade policies may apply to a given server.  The
+upgrade policy which maps to the most recent teleport version will be given priority for
+a given server.
+
+Policy evaluation (i.e. determining how releases map to servers) will be the responsibility
+of the auth server.  The goal of this centralization is to ensure that if a logic bug in the
+mapping results in requiring a manual upgrade, only the auth server(s) will need to be upgraded.
+Some amount of hedging will be added to the API to make sure its easy to add custom (external)
+policy evaluators later if the need exists, but I imagine that custom upgrade controllers and
+manually managed release manifests will cover most usecases.
+
+I have yet to work out the exact details of how policy syntax and evaluation should work, but
+here's an example of one of the syntaxes I've been toying with in my head:
+
+```yaml
+target_selectors:
+  - name: Latest oss
+    flavors: [oss]
+    permit_unstable: true
+server_selectors:
+  - name: Preview servers
+    server_roles: ['proxy','node']
+    filter: 'contains(server.labels['unstable-preview'],'yes')'
+schedule:
+  not_before: '2021-04-29T00:00:00Z'
+  time_range: '01:00:00-06:00:00'
+  day_range: Mon-Fri
+```
+
+Note that my current thinking on target/server matching is that policies should not actually
+be able to cause a node to be upgraded to a target with different fundamental attributes (e.g. os,
+flavor, arch, etc) due to the risk of catastrophic user error being too great. Instead, the
+purpose of target selectors it to allow different requirements to be put in place *based* on
+those fundamental attributes (e.g. only installing unstable versions on oss nodes).
+
+Not all policies need be fancy.  In fact, something like this aught to be a valid policy:
+
+```
+target_selectors:
+- version: '99.*'
+```
+
+read: "Keep all servers on the latest stable `99.X` release"
+
+
+#### Upgrade Controllers
+
+The "upgrade controller" will be the primary hook that is used to customize upgrade
```

Comment:
* What controllers are we going to ship first?

I assume we would have 2 controllers:

* Deb/Yum controller that devs can customize
* Builtin controller?

* What's our UX for the administrator of the controller?

For example, the server triggered an update of a cluster of 5 nodes. Questions they would have:

* How is update going?
* If the server has failed, initiate a rollback.
* Go to the server and manually fix the problem in case if the system works?

* How would commands look like?

Second problem is where to keep update state. We can assume that database is where we store upgrade state, things become much simpler - we assume that the database is our source of truth about cluster desired state, just like with Kubernetes.

Reply:
My current thinking is that we ship with just the `script` controller.  Installation via `deb`/`yum` are examples of how the script controller can be used, but are not themselves controllers.  I think that *for now* the default should be something very simple that pulls from `get.gravitational.com`.  I'm thinking that we circumvent the dependency on utilities like `b2sum` by providing purpose-specific hidden `teleport` subcommands that perform similar checks for the time being.

Added an upgrade UX section in latest commit which should clarify my thinking on UX and how status is published.